### DISCUSSION TITLE
Announcing Sanity Studio v3: Developer Preview

### DISCUSSION THREAD

**Discussion by kmelve** [sanity-io/sanity#3327] [Announcements]
The [Developer Preview of Sanity Studio Version 3](https://www.sanity.io/blog/sanity-studio-v3-developer-preview) is now available for you to try!

We’ve significantly upgraded the developer experience, making it faster and easier to customize the studio to your needs.

Here are a few of the highlights:

* A new bundler agnostic architecture that makes it possible to embed the Studio inside other applications
* New and improved APIs for configuration, form customization, plugins, and themes (coming soon)
* Workspaces: Easily switch between projects, datasets, and configurations within the same Studio application
* Studio APIs are fully typed with TypeScript and come with new reference API documentation (in beta)
* Plugin toolkit to support development, publishing, and migrations
* Dark mode support

Heads up: Sanity Studio Version 3 introduces breaking changes and requires manual code migration – mainly where “parts” are used (`sanity.json` and `import X from “part:X`).

We know this can cause some disruption, but we’re here to support you along the way! We’ve published migration guides and will actively take feedback and answer questions on GitHub. We’ve also taken our official plugins and made them v3-ready. As we move towards general availability, we’re dedicated to ensuring that your migration experience is as smooth as possible.

Next steps:

* [Read the release blog post](http://sanity.io/blog/sanity-studio-v3-developer-preview)
* [Explore the new Sanity Docs Beta](https://beta.sanity.io/docs)
* [Read Best Practices for Positive & Empathetic Discussions](https://github.com/sanity-io/sanity/discussions/3333)
* [Join the conversation here on GitHub](https://github.com/sanity-io/sanity/discussions)

We can’t wait to answer your questions and hear what you think!

**Comment by wrux** [comment]
I'm mid-way through migrating a Studio instance, but hit a roadblock on getting the `@sanity/dashboard` and `@sanity/vision` plugins setup. Will all first party Sanity plugins be supported by V3 in the beta stages?

**Reply by p10e** [reply]
Great question, @wrux! If you're available, I recommend joining the [launch event](https://www.meetup.com/meetup-group-dvjyrjdv/events/286293446/) today, where there will be a section on migrating tools and plugins. Some of the packages will be available on npm under `@studio-v3` tags today, others will follow in a few days, and again others will come a bit later.

V3 of `@sanity/dashboard` plugin should be available under `@sanity/dashboard@studio-v3` today (readme [here](https://www.npmjs.com/package/@sanity/dashboard/v/3.0.0-v3-studio.0)). V3 of `@sanity/vision` will not be available just yet but we will keep you posted.

**Comment by vincentjflorio** [comment]
Is this related to the package or is this me and I should delete this message? Windows 10, VS Code terminal, clean cache and empty directory.

```
node:events:504
      throw er; // Unhandled 'error' event
      ^

Error: spawn C:\Users\thevi\AppData\Local\npm-cache\_npx\681b0a19326c4f58\node_modules\@sanity\cli\bin\sanity ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:283:19)
    at onErrorNT (node:internal/child_process:478:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (node:internal/child_process:289:12)
    at onErrorNT (node:internal/child_process:478:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn C:\\Users\\thevi\\AppData\\Local\\npm-cache\\_npx\\681b0a19326c4f58\\node_modules\\@sanity\\cli\\bin\\sanity',
  path: 'C:\\Users\\thevi\\AppData\\Local\\npm-cache\\_npx\\681b0a19326c4f58\\node_modules\\@sanity\\cli\\bin\\sanity',
  spawnargs: [ 'init' ]
}
npm ERR! code 1
npm ERR! path C:\Users\thevi\OneDrive\Documents\GitHub\sanity-v3
npm ERR! command failed
npm ERR! command C:\windows\system32\cmd.exe /d /s /c create-sanity
```

**Reply by p10e** [reply]
Thanks for sharing, @vincentjflorio! To rule out one case where you might see this `ENOENT` error, could you try setting your path variable as described in [this guide](https://www.computerhope.com/issues/ch000549.htm#windows10)? If that does not improve the situation, we'll dig deeper!

**Reply by vincentjflorio** [reply]
Thanks for getting back with me so quickly! I am excited to start.

My path variable is set, though perhaps improperly. I tried editing and removing pnpm's entry in case it was complicating things, and fully exited everything node- or command line-related before trying again each time.

If it's helpful, this is the only package I have had an issue with like this, and I even tried yarn and eventually got the same result after resolving an esbuild dependency issue. 

That said, I don't know what I don't know so I am reproducing the current path just in case something is garbled that I don't even know enough to be unaware of to tell you =)

`C:\Users\thevi\AppData\Roaming\npm;C:\Users\thevi\AppData\Local\Programs\Python\Python39\;C:\Users\thevi\AppData\Local\Microsoft\WindowsApps;C:\Program Files\heroku\bin;C:\Users\thevi\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\thevi\AppData\Roaming\Composer\vendor\bin;%JAVA_HOME%\bin;C:\Users\thevi\AppData\Local\GitHubDesktop\bin;C:\Program Files\PostgreSQL\14\bin\;`

**Reply by frozenskys** [reply]
I'm seeing this as well on Windows 10 - path looks fine, empty folder, latest sanity cli installed along with node v16.15.1 and npm version 8.12.1. npm log here: [2022-06-14T22_19_32_605Z-debug-0.log](https://github.com/sanity-io/sanity/files/8904292/2022-06-14T22_19_32_605Z-debug-0.log)

**Reply by vincentjflorio** [reply]
We might have a smidge of company on #3336 and #3337

**Reply by tbassetto** [reply]
Same issue on Windows 10, node 16.15.0 and npm 8.5.5. The questions is: has someone been able to use the dev preview on Windows 10?

```
➜ npm create sanity@dev-preview
Need to install the following packages:
  create-sanity@dev-preview
Ok to proceed? (y)
node:events:505
      throw er; // Unhandled 'error' event
      ^

Error: spawn C:\Users\THOBAS\AppData\Local\npm-cache\_npx\681b0a19326c4f58\node_modules\@sanity\cli\bin\sanity ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:283:19)
    at onErrorNT (node:internal/child_process:478:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (node:internal/child_process:289:12)
    at onErrorNT (node:internal/child_process:478:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn C:\\Users\\THOBAS\\AppData\\Local\\npm-cache\\_npx\\681b0a19326c4f58\\node_modules\\@sanity\\cli\\bin\\sanity',
  path: 'C:\\Users\\THOBAS\\AppData\\Local\\npm-cache\\_npx\\681b0a19326c4f58\\node_modules\\@sanity\\cli\\bin\\sanity',  spawnargs: [ 'init' ]
}
npm ERR! code 1
npm ERR! path C:\Users\THOBAS\Code\sanity-v3
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c create-sanity

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\THOBAS\AppData\Local\npm-cache\_logs\2022-06-15T06_38_10_452Z-debug-0.log
```

**Reply by wiwalla** [reply]
It looks like there is the same issue on Windows 11 too. I get the same exact error. 
Could it be the double "\\" in the path and syscall that makes the command fail?

**Reply by p10e** [reply]
Thanks all for extensive reporting with logs and version numbers - very helpful. Sorry this issue is preventing you from creating a v3 studio right now. We're investigating why it might be happening on Windows 10/11 in particular.

May we ask you to try running the command below to see if it still throws the error?

```npx @sanity/cli@dev-preview init```

**Reply by wiwalla** [reply]
The  `npx @sanity/cli@dev-preview init ` works but it looks like it just makes V2. (The sanity.json file gets created instead of the sanity.config.js)

 Here are the version results
`@sanity/cli                 2.30.0 (up to date)
@sanity/base                  2.30.1 (up to date)
@sanity/core                  2.30.1 (up to date)
@sanity/default-layout        2.30.1 (up to date)
@sanity/default-login         2.30.1 (up to date)
@sanity/desk-tool             2.30.1 (up to date)
@sanity/eslint-config-studio   2.0.0 (up to date)
@sanity/vision                2.30.1 (up to date)`

**Reply by tremppu** [reply]
Had same issues with Windows 11 but got it working by: `npm i -g @sanity/cli@dev-preview` and then just `sanity init`. 

But then got this (in browser, Chrome v102) when tried to run studio:

`Uncaught error: url.parse is not a function
http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:152927:5
TypeError: url.parse is not a function
    at module2.exports (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:37092:22)
    at module2.exports (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:37226:42)
    at http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:37418:28
    at Object.publish3 [as publish] (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:36548:27)
    at Observable3._subscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:37533:30)
    at Observable3._trySubscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:34316:23)
    at Observable3.subscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:34302:148)
    at Observable3._subscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:38212:72)
    at Observable3._trySubscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:34316:23)
    at Observable3.subscribe (http://localhost:3333/node_modules/.vite/deps/chunk-QIL6IAPE.js?v=0aebf5c3:34302:148)
`

**Reply by wiwalla** [reply]
> 

Thanks for this tip. I got it to work now. 
This error came up for me too. I just did a `npm install url` and then I could log on to sanity and see my sanity studio!

**Reply by vincentjflorio** [reply]
> > 
> 
> Thanks for this tip. I got it to work now. This error came up for me too. I just did a `npm install url` and then I could log on to sanity and see my sanity studio!

These steps got me set up just fine on Windows 10; the @tremppu suggestion of the "npm i" got me started, followed by having to install the url package.

**Reply by frozenskys** [reply]
So if I remove my global sanity (v2.30.0) using `npm uninstall -g @sanity/cli` then run `npx @sanity/cli@dev-preview init` it will create me a nice shiny V3 studio which is missing the url package. Running a `npm i url` in the studio folder fixes this, and all appears well with the world.

If however I leave the sanity/cli v2 installed, then running `npx @sanity/cli@dev-preview init` asks to install sanity/cli v3.0.0 but actually creates a v2 studio install 😬 

Also I'm assuming to choose `pnpm` as the package manager for the v3 install (as per the instructions) I would already need to have installed pnpm - as it does not appear as an option?

**Comment by mylesthedev** [comment]
Is there a rough date set for general availability? i.e. production ready?

**Reply by p10e** [reply]
Hi @mediabeastnz, I'm afraid we don't have a rough date for GA just yet. Although we want it to be out as soon as possible, the goal of the Developer Preview release is to get feedback, learn, and iterate. While we follow up on those outcomes, we'll be able to narrow down on a release date. We'll make sure to keep you posted!

**Reply by adekoelijah** [reply]
Please I need help for sanity installation, it keep saying error.

Please help, Thanks

**Comment by raulfdm** [comment]
I think the integration with `pnpm` isn't working properly.

I've:

1. created a new sanity project with the command `npm create sanity@dev-preview`
    - TypeScript
    - Default Config
    - Blog start
    - Install deps with pnpm
1. tried to run `pnpm run start`

This throw the following error:

```
> sanity-v3@1.0.0 start /Users/raulmelo/development/sandbox/sanity-v3
> sanity start

/Users/raulmelo/development/sandbox/sanity-v3/node_modules/.pnpm/sanity@3.0.0-dev-preview.8_zxhai4qal7ezfht52cx4u672cu/node_modules/sanity/node_modules/.bin/sanity:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1033:15)
    at Module._compile (node:internal/modules/cjs/loader:1069:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Module._load (node:internal/modules/cjs/loader:827:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/raulmelo/development/sandbox/sanity-v3/node_modules/.pnpm/sanity@3.0.0-dev-preview.8_zxhai4qal7ezfht52cx4u672cu/node_modules/sanity/bin/sanity:10:3)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
```

## Environment

- pnpm version: 7.2.1
- npm version: 8.5.5
- node version: 18.3.0
- OS
    - Platform: darwin
    - Arch: arm64
    - Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000

## Relevant gotchas

I have a gut feeling that's related to how the dependencies are installed.

When I installed the deps, I got errors about missing peer dependencies, but have installed without them ([more on that here](https://pnpm.io/errors#err_pnpm_peer_dep_issues)).

When I tried to set [`auto-install-peers=true`](https://pnpm.io/npmrc#auto-install-peers) in a `.npmrc` file so pnpm can force install those missing dependencies, the error is still there.

If I tried set [`shamefully-hoist=true`](https://pnpm.io/npmrc#shamefully-hoist) in a `.npmrc` file so pnpm can uses the npm/yarn installation strategy, the error is still there.

If I remove `pnpm-lock.yaml` and the `node_modules` and use `npm` instead (installing and running start), **the error is gone**.

I'm not sure if reporting this error here is relevant, if not, I can open an issue with some demos.

**Comment by IPlayJazz** [comment]
I finished moving my project to V3.
Is there already a V3 release date?

**Reply by quirm** [reply]
It would be great to see a committed date. The changes and the direction of V3 are fabulous but the amount of bugs for the fundamental user-facing Studio functionality makes me worry if that date is in months or years.

**Reply by wrux** [reply]
@quirm A v3 release candidate was published recently, so I assume the official release is imminent
